src/Microsoft.Data.Analysis/DataFrameWindow.Rolling.PrimitiveDataFrameColumn.cs

```diff
@@ -0,0 +1,213 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Microsoft.Data.Analysis
+{
+    /// <summary>
+    /// A window for <see cref="PrimitiveDataFrameColumn{T}"/> to support rolling window operations
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public class PrimitiveDataFrameColumnRollingWindow<T> : DataFrameColumnWindow
+        where T : unmanaged
+    {
+        private int _windowSize;
+        private PrimitiveDataFrameColumn<T> _currentColumn;
+        private long _currentIndex;
+
+        internal PrimitiveDataFrameColumnRollingWindow(int windowSize, PrimitiveDataFrameColumn<T> currentColumn)
+        {
+            _windowSize = windowSize;
+            _currentColumn = currentColumn;
+        }
+
+        /// <summary>
+        /// Apply <paramref name="func"/> over a rolling window of values. 
+        /// </summary>
+        /// <typeparam name="U">The returned result for each rolling window</typeparam>
+        /// <param name="func">The delegate to apply for each window. The first parameter is a linked list of values in the current window. The second parameter is the current index in the column. The current window ends at the current index and include values preceeding it</param>
+        /// <returns>The result computed for each rolling window</returns>
+        public PrimitiveDataFrameColumn<U> Apply<U>(Func<LinkedList<T?>, long, U?> func)
+            where U : unmanaged
+        {
+            return _currentColumn.ApplyRollingFunc(func, _windowSize);
+        }
+
+        /// <summary>
+        /// Finds the min of the values in this window
+        /// </summary>
+        /// <returns>Returns a DataFrame containing the min in each window</returns>
+        public new PrimitiveDataFrameColumn<T> Min()
+        {
+            _currentIndex = 0;
+
+            LinkedList<KeyValuePair<double, long>> sortedList = new LinkedList<KeyValuePair<double, long>>();
+            LinkedList<T?> windowValues = new LinkedList<T?>();
+            IDoubleConverter<T> doubleConverter = DoubleConverter<T>.Instance;
+            return _currentColumn.Apply<T>((T? value) =>
+            {
+                double doubleValue = value.HasValue ? doubleConverter.GetDouble(value.Value) : double.MaxValue;
+                // This algorithm builds a sorted list. Time complexity is O(N) here since each value is added/removed to sortedList only once.
+                // If the values at the end of the sortedList are greater than value, delete them since they can never be the min in this window
+                // If value is null, treat it as double.MaxValue
+                if (value.HasValue)
```

Comment:
Is this needed as `doubleValue` already does this distinction? I.e. it is either the value or `double.MaxValue`, so this `if` can be omitted and both `while`-loops can be "merged" to one (use the first one and it's good).

Reply:
Yup, my bad. I added the `doubleValue` at the end and didn't realize this could be cleaned up.